torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -935,27 +935,34 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
   }
 }
 
-void ProcessGroupNCCL::logWorkStart(WorkNCCL& work) {
-  if (work.startTraceUpdated_)
+void ProcessGroupNCCL::logWorkStart(WorkNCCL& work, bool emitDesyncInfo) {
+  if (terminateProcessGroup_.load() || work.startTraceUpdated_)
     return;
+  work.startTraceUpdated_ = true;
+
+  emitCollectiveStart(work);
 
-  if (terminateProcessGroup_.load() || storeError_)
+  if (!emitDesyncInfo || storeError_)
     return;
 
-  work.startTraceUpdated_ = true;
   storeError_ = !c10d::traceUpdate(
       store_, traceKeyStart_, work.seq_, opTypeToString(work.opType_));
 }
 
-void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
-  if (terminateProcessGroup_.load() || storeError_)
+void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work, bool emitDesyncInfo) {
+  if (terminateProcessGroup_.load())
     return;
 
   // In case the start of the work hasn't been logged
```

Comment:
do we have cases where we would not have called 'emitCollectiveStart' for this work? i don't understand the reason for `!work.startTraceUpdated_`